MIGRATING.md

```diff
@@ -0,0 +1,140 @@
+# Migrating from qiskit-ibmq-provider>=0.16
+
+## Installation
+The Qiskit IBM Provider is now distributed as a separate PyPI package called `qiskit-ibm`. You can install the provider using pip:
+
+```bash
+pip install qiskit-ibm
+```
+
+## Breaking Changes
+1. The `IBMQ` global variable which was an instance of the `IBMQFactory` has been removed.
+2. `IBMQFactory` has been renamed to `IBMAccount` class.
+3. `AccountProvider` class has been renamed to `IBMProvider` class.
+
+    For example, if you are looking to quickly migrate your existing code:
+
+    Before
+    ```python
+    from qiskit import IBMQ
+    IBMQ.load_account()
+    provider = IBMQ.get_provider(hub='ibm-q', group='test', project='default')
+    simulator_backend = provider.get_backend('ibmq_qasm_simulator')
+    ```
+    After
+    ```python
+    from qiskit_ibm import IBMAccount
+    account = IBMAccount()
+    account.load_account() # or account.load()
+    provider = account.get_provider(hub='ibm-q', group='test', project='default') # or account.provider(...)
+    simulator_backend = provider.get_backend('ibmq_qasm_simulator')
+    ```
+
+    For any new code you are strongly encouraged to directly use the `IBMProvider` class which provides a simplified interface and is atleast 50% faster than instantiating via the `IBMAccount` class when you are working mostly with a single project (hub/group/project). For example above program will look like:
```

Comment:
Yes IBMProvider is faster because it doesn’t initialize all providers and in turn all back ends when a user has more than one provider. But that is only an added benefit.

The primary goal is a simple interface with IBMProvider as an entry point. (the naming of the entry point I was told was non negotiable) I also learnt that we have to start moving towards a design like in other providers and that we will loose the concept of hgp when we move to the IBM cloud.

With all this in mind making IBMAccount work like you have shown in the examples is not an option since the primary entry point should be IBMProvider. One should use IBMAccount just to save credentials and then use IBMProvider in most snippets. That’s my thought here.

Therefore IBMProvider will match the design of the other providers we have wherein the primary entry point represents a single account + hgp.

Having it like IBMProvider and AccountProvider is even more confusing since we are now calling our Python library as IBMProvider and then we have another concept of provider as in HGP. I am not completely against this and I am open to doing this if you can run it by business and make sure this is ok since we are deviating from the design of other partner providers.

Reply:
I'm not against having `IBMProvider` as the entry point. I'm, however, against having two entry points (this PR currently allows both `IBMAccount().provider()` or `IBMProvider()`). 

>  One should use IBMAccount just to save credentials and then use IBMProvider in most snippets. That’s my thought here.

If `IBMAccount` is only used to manage credentials, then it shouldn't have the `provider()` method. Furthermore, I'd argue that other providers "manage" credentials in the `Provider` constructor by making the user pass token/url. They don't use a separate class for that. 

>  I also learnt that we have to start moving towards a design like in other providers and that we will loose the concept of hgp when we move to the IBM cloud.

> Having it like IBMProvider and AccountProvider is even more confusing since we are now calling our Python library as IBMProvider and then we have another concept of provider as in HGP

It has always been confusing since we are the only "provider" that has multiple "sub-providers" (hgp). If `AccountProvider` is too confusing, we can just give it another name, perhaps `IBMSubProvider` or `IBMHubGroupProject` (feel free to suggest better names)? The name is not really all that important since it's mostly hidden from the user.

So I think my previous proposal is still reasonable, that users can do 
```
provider = IBMProvider()  # same as IBMQ.load_account()
provider = IBMProvider(token=foo, url=bar)  # same as IBMQ.enable_account(foo, bar)
provider = IBMProvider(project=project)  # same as load_account() + get_provider()
provider = IBMProvider(token=foo, save=True)  # same as save_account() + get_provider()
```

Then when we are ready to transition out of hgp, we can just deprecate the `hub`, `group`, `project` keywords. 

@nonhermitian any thoughts on this?

Reply:
> I'm not against having `IBMProvider` as the entry point. I'm, however, against having two entry points (this PR currently allows both `IBMAccount().provider()` or `IBMProvider()`).

It'd be nice to have just one entry point `IBMProvider`. I agree.

> If `IBMAccount` is only used to manage credentials, then it shouldn't have the `provider()` method. Furthermore, I'd argue that other providers "manage" credentials in the `Provider` constructor by making the user pass token/url. They don't use a separate class for that.

The AQT Provider has two entry points. https://github.com/Qiskit-Partners/qiskit-aqt-provider
So I want to know if it is ok and if it makes sense to deviate from this design. I have some questions about using `Provider` constructor to save credentials, see below.

> It has always been confusing since we are the only "provider" that has multiple "sub-providers" (hgp). If `AccountProvider` is too confusing, we can just give it another name, perhaps `IBMSubProvider` or `IBMHubGroupProject` (feel free to suggest better names)? The name is not really all that important since it's mostly hidden from the user.

Yes not really important if we go with your below proposal, but I would not include the word `Provider` in the class name when renaming `AccountProvider`, since it becomes difficult to explain the two different provider concepts in the documentation. Also this name shows up to the user when we get providers. For example right now it shows up like:
```<AccountProvider for IBMQ(hub='ibm-q', group='open', project='main>```

I would suggest using `IBMProject` or `IBMHubGroupProject`.

> So I think my previous proposal is still reasonable, that users can do
> 
> ```
> provider = IBMProvider()  # same as IBMQ.load_account()
> provider = IBMProvider(token=foo, url=bar)  # same as IBMQ.enable_account(foo, bar)
> provider = IBMProvider(project=project)  # same as load_account() + get_provider()
> provider = IBMProvider(token=foo, save=True)  # same as save_account() + get_provider()
> ```
> 
> Then when we are ready to transition out of hgp, we can just deprecate the `hub`, `group`, `project` keywords.

Few questions:
1. Let's say I have a snippet in a cell like this
```
from qiskit_ibm import IBMProvider
provider = IBMProvider(token=foo, save=True)
backend = provider.backend('ibmq_qasm_simulator`)
backend.run(...)
```
what happens if I run my snippet again? Are my credentials saved again?

2. How does a user see the list of their providers (h/g/p)?
```
from qiskit_ibm import IBMProvider
provider = IBMProvider()
provider.providers() or provider.projects()? # This doesn't make sense since `provider` variable could point to a single hgp already
```
3. Also how does one delete a saved account?
```
from qiskit_ibm import IBMProvider
provider = IBMProvider()
provider.delete() ?
```

Reply:
> The AQT Provider has two entry points. https://github.com/Qiskit-Partners/qiskit-aqt-provider

Yes but `AQTAccount()` is [deprecated](https://github.com/Qiskit-Partners/qiskit-aqt-provider/blob/master/qiskit_aqt_provider/aqt_account.py#L30).

> Yes not really important if we go with your below proposal, but I would not include the word Provider in the class name when renaming AccountProvider

I was wrong about having to rename `AccountProvider`, since `IBMProvider` _is_ the new name of `AccountProvider`, with some changes. When one does `provider = IBMProvider()`, they are getting back a single h/g/p "provider" already.

> 1. Let's say I have a snippet in a cell like this
> 
> ```
> from qiskit_ibm import IBMProvider
> provider = IBMProvider(token=foo, save=True)
> backend = provider.backend('ibmq_qasm_simulator`)
> backend.run(...)
> ```
> 
> what happens if I run my snippet again? Are my credentials saved again?

Yes. We can add an `overwrite` parameter, but I don't think it's necessary.

> 
> 1. How does a user see the list of their providers (h/g/p)?
> 
> ```
> from qiskit_ibm import IBMProvider
> provider = IBMProvider()
> provider.providers() or provider.projects()? # This doesn't make sense since `provider` variable could point to a single hgp already
> ```

We can perhaps make it a class method, like `IBMProvider.list()`

> 
> 1. Also how does one delete a saved account?
> 
> ```
> from qiskit_ibm import IBMProvider
> provider = IBMProvider()
> provider.delete() ?
> ```

Same as above, especially since `delete_account()` is already a static method. 

Some other alternatives can be 
a) make account management functions and not part of any class. I don't like this because functions are hard to find.
b) have a separate `IBMAccount` that does just account management (i.e. no `get_provider()`). But most will be static/class methods - e.g. `IBMAccount.save()`, so is it really better than `IBMProvider.save_account()`?
c) same as (b) but through `IBMProvider` - e.g. `provider.account().save()`.

Reply:
> Yes but AQTAccount() is deprecated.

Oops didn't notice that. That's great then!

> Some other alternatives can be
a) make account management methods individual functions and not part of any class. I don't like this because functions are hard to find.
b) have a separate IBMAccount that does just account management (i.e. no get_provider()). But most (all?) will be static/class methods - e.g. IBMAccount.save(), so is it really better than IBMProvider.save_account()?
c) same as (b) but through IBMProvider - e.g. provider.account().save().

Ok awesome! I like the idea of having static/class methods for account management on the `IBMProvider` class itself, so we can have one entry point.

Reply:
Fixed!